python/pyiceberg/transforms.py

```diff
@@ -511,6 +590,31 @@ def preserves_order(self) -> bool:
     def source_type(self) -> IcebergType:
         return self._source_type
 
+    def project(self, name: str, pred: BoundPredicate) -> Optional[UnboundPredicate]:
+        field_type = pred.term.ref().field.field_type
+
+        if isinstance(pred.term, BoundTransform):
+            return _project_transform_predicate(self, name, pred)
+
+        # Implement startswith and notstartswith for string (and probably binary)
+        # https://github.com/apache/iceberg/issues/6112
+
+        if isinstance(pred, BoundUnaryPredicate):
+            return pred.as_unbound(Reference(name))
+
+        if isinstance(field_type, (IntegerType, LongType, DecimalType)):
+            if isinstance(pred, BoundLiteralPredicate):
+                return _truncate_number(name, pred, self.transform(field_type))
+            elif isinstance(pred, BoundIn):
+                return _transform_set(name, pred, self.transform(field_type))
```

Comment:
Minor: this case could be moved to the higher level since it is the same for both numeric and array types.

Reply:
I like it 👍🏻

Reply:
Actually, `_truncate_number` and `_truncate_array` is different

Reply:
Yeah, I just meant for the inner `elif` case, where both call `_set_apply_transform`.